File: sdk/runanywhere-swift/Sources/RunAnywhere/Core/Models/Configuration/ModelDownloadConfiguration.swift
import Foundation

// MARK: - Download Policy

/// Download policy for models
public enum DownloadPolicy: String, Codable, Sendable {
    case automatic = "automatic"      // Download automatically if needed
    case wifiOnly = "wifi_only"       // Only download on WiFi
    case manual = "manual"             // Require user confirmation
    case never = "never"               // Don't download, fail if not available
}

// MARK: - Model Download Configuration

/// Simple configuration for model downloads
public struct ModelDownloadConfiguration: Codable, Sendable {

    /// Download policy
    public var policy: DownloadPolicy

    /// Maximum concurrent downloads
    public var maxConcurrentDownloads: Int

    /// Number of retry attempts
    public var retryCount: Int

    /// Download timeout in seconds
    public var timeout: TimeInterval

    /// Enable background downloads
    public var enableBackgroundDownloads: Bool

    public init(
        policy: DownloadPolicy = .automatic,
        maxConcurrentDownloads: Int = 3,
        retryCount: Int = 3,
        timeout: TimeInterval = 300.0,
        enableBackgroundDownloads: Bool = false
    ) {
        self.policy = policy
        self.maxConcurrentDownloads = maxConcurrentDownloads
        self.retryCount = retryCount
        self.timeout = timeout
        self.enableBackgroundDownloads = enableBackgroundDownloads
    }
}

// MARK: - Simple Helper

extension ModelDownloadConfiguration {

    /// Check if download is allowed
    public func shouldAllowDownload(isWiFi: Bool = false, userConfirmed: Bool = false) -> Bool {
        switch policy {
        case .automatic:
            return true
        case .wifiOnly:
            return isWiFi
        case .manual:
            return userConfirmed
        case .never:
            return false
        }
    }
}

File: sdk/runanywhere-flutter/lib/core/models/configuration/model_download_configuration.dart
import 'download_policy.dart';

/// Simple configuration for model downloads.
/// Matches iOS ModelDownloadConfiguration from Configuration/ModelDownloadConfiguration.swift
class ModelDownloadConfiguration {
  /// Download policy
  final DownloadPolicy policy;

  /// Maximum concurrent downloads
  final int maxConcurrentDownloads;

  /// Number of retry attempts
  final int retryCount;

  /// Download timeout in seconds
  final Duration timeout;

  /// Enable background downloads
  final bool enableBackgroundDownloads;

  const ModelDownloadConfiguration({
    this.policy = DownloadPolicy.automatic,
    this.maxConcurrentDownloads = 3,
    this.retryCount = 3,
    this.timeout = const Duration(seconds: 300),
    this.enableBackgroundDownloads = false,
  });

  /// Create from JSON map
  factory ModelDownloadConfiguration.fromJson(Map<String, dynamic> json) {
    return ModelDownloadConfiguration(
      policy: DownloadPolicy.fromRawValue(
              json['policy'] as String? ?? 'automatic') ??
          DownloadPolicy.automatic,
      maxConcurrentDownloads:
          (json['maxConcurrentDownloads'] as num?)?.toInt() ?? 3,
      retryCount: (json['retryCount'] as num?)?.toInt() ?? 3,
      timeout: Duration(seconds: (json['timeout'] as num?)?.toInt() ?? 300),
      enableBackgroundDownloads:
          json['enableBackgroundDownloads'] as bool? ?? false,
    );
  }

  /// Convert to JSON map
  Map<String, dynamic> toJson() {
    return {
      'policy': policy.rawValue,
      'maxConcurrentDownloads': maxConcurrentDownloads,
      'retryCount': retryCount,
      'timeout': timeout.inSeconds,
      'enableBackgroundDownloads': enableBackgroundDownloads,
    };
  }

  /// Create a copy with updated fields
  ModelDownloadConfiguration copyWith({
    DownloadPolicy? policy,
    int? maxConcurrentDownloads,
    int? retryCount,
    Duration? timeout,
    bool? enableBackgroundDownloads,
  }) {
    return ModelDownloadConfiguration(
      policy: policy ?? this.policy,
      maxConcurrentDownloads:
          maxConcurrentDownloads ?? this.maxConcurrentDownloads,
      retryCount: retryCount ?? this.retryCount,
      timeout: timeout ?? this.timeout,
      enableBackgroundDownloads:
          enableBackgroundDownloads ?? this.enableBackgroundDownloads,
    );
  }

  /// Check if download is allowed based on policy
  bool shouldAllowDownload({bool isWiFi = false, bool userConfirmed = false}) {
    switch (policy) {
      case DownloadPolicy.automatic:
        return true;
      case DownloadPolicy.wifiOnly:
        return isWiFi;
      case DownloadPolicy.manual:
        return userConfirmed;
      case DownloadPolicy.never:
        return false;
    }
  }
}
